File: backend/src/web/one_drive_sync.py
import os
import msal
import shutil
import requests

from typing import Callable, Dict, Any
from ..utils.definitions import GRAPH_API_ENDPOINT


class OneDriveSync:
    """
    Manages synchronization for OneDrive using MS Graph API and MSAL.
    """
    def __init__(
        self,
        local_source_path: str,
        drive_destination_folder_name: str,
        client_id: str = None,
        client_secret: str = None, # Not typically used for Public Client (Desktop App) flow
        dry_run: bool = False,
        logger: Callable[[str], None] = print,
        action_local_orphans: str = "upload",
        action_remote_orphans: str = "download"
    ):
        self.local_path = local_source_path
        self.remote_path = drive_destination_folder_name.strip("/")
        self.client_id = client_id
        self.dry_run = dry_run
        self.logger = logger
        self.action_local = action_local_orphans
        self.action_remote = action_remote_orphans
        self._is_running = True
        
        self.access_token = None
        self.headers = None
        self.remote_path_to_id = {} # Cache for folder IDs

    def check_stop(self):
        if not self._is_running:
            raise InterruptedError("Synchronization manually interrupted.")

    def _authenticate(self):
        """
        Authenticates using MSAL Device Code Flow (safest for desktop apps without needing a redirect URI).
        """
        if not self.client_id:
            raise ValueError("OneDrive Client ID is missing.")

        scopes = ["Files.ReadWrite.All", "User.Read"]
        app = msal.PublicClientApplication(self.client_id)

        self.logger("🔑 Authenticating with Microsoft Graph...")
        
        # 1. Try silent token acquisition (cached)
        accounts = app.get_accounts()
        result = None
        if accounts:
            result = app.acquire_token_silent(scopes, account=accounts[0])

        # 2. If silent fails, use Device Code Flow
        if not result:
            self.logger("   Initiating Device Code Flow...")
            flow = app.initiate_device_flow(scopes=scopes)
            if "user_code" not in flow:
                raise ValueError("Failed to create device flow.")

            # Show code to user in logs (Ideally this would be a popup, but logs work for now)
            self.logger(f"\n⚠️  ACTION REQUIRED ⚠️")
            self.logger(f"   1. Open: {flow['verification_uri']}")
            self.logger(f"   2. Enter Code: {flow['user_code']}")
            self.logger(f"   Waiting for authentication...\n")

            result = app.acquire_token_by_device_flow(flow)

        if "access_token" in result:
            self.access_token = result["access_token"]
            self.headers = {'Authorization': 'Bearer ' + self.access_token}
            self.logger("✅ Authentication successful.")
        else:
            error = result.get("error")
            desc = result.get("error_description")
            raise RuntimeError(f"Authentication failed: {error} - {desc}")

    def _get_local_files_map(self) -> Dict[str, Dict[str, Any]]:
        """Creates a map of relative_path -> metadata for local files."""
        local_items = {}
        base_len = len(self.local_path) + len(os.sep) 

        for root, dirs, files in os.walk(self.local_path):
            self.check_stop()
            for file in files:
                abs_path = os.path.join(root, file)
                rel_path = abs_path[base_len:].replace(os.sep, '/')
                local_items[rel_path] = {'path': abs_path, 'is_folder': False}
            for d in dirs:
                abs_path = os.path.join(root, d)
                rel_path = abs_path[base_len:].replace(os.sep, '/')
                local_items[rel_path] = {'path': abs_path, 'is_folder': True}
        return local_items

    def _get_remote_files_map(self) -> Dict[str, Dict[str, Any]]:
        """
        Recursively lists files from OneDrive.
        Since Graph API doesn't have a simple "recursive list" for a specific folder path easily accessible,
        we traverse layer by layer.
        """
        self.logger(f"🔍 Scanning remote folder: {self.remote_path}")
        remote_items = {}
        
        # Resolve root folder ID
        root_id = self._resolve_path_to_id(self.remote_path)
        if not root_id:
            self.logger("   Remote folder does not exist yet.")
            return {}

        # Queue for traversal: (folder_id, relative_path_prefix)
        queue = [(root_id, "")]

        while queue:
            self.check_stop()
            current_id, current_rel = queue.pop(0)
            
            url = f"{GRAPH_API_ENDPOINT}/me/drive/items/{current_id}/children"
            
            while url:
                self.check_stop()
                response = requests.get(url, headers=self.headers)
                if response.status_code != 200:
                    self.logger(f"❌ Error listing children for {current_rel}: {response.text}")
                    break
                
                data = response.json()
                items = data.get('value', [])
                
                for item in items:
                    name = item['name']
                    item_id = item['id']
                    is_folder = 'folder' in item
                    rel_path = os.path.join(current_rel, name).replace(os.sep, '/') if current_rel else name
                    
                    remote_items[rel_path] = {
                        'id': item_id,
                        'is_folder': is_folder
                    }
                    
                    if is_folder:
                        queue.append((item_id, rel_path))
                        # Cache ID for potential uploads later
                        full_path_key = f"{self.remote_path}/{rel_path}" if self.remote_path else rel_path
                        self.remote_path_to_id[full_path_key] = item_id

                url = data.get('@odata.nextLink') # Pagination

        return remote_items

    def _resolve_path_to_id(self, path: str):
        """Resolves a remote path to an Item ID."""
        if not path: return "root"
        
        # Check cache
        if path in self.remote_path_to_id:
            return self.remote_path_to_id[path]

        url = f"{GRAPH_API_ENDPOINT}/me/drive/root:/{path}"
        response = requests.get(url, headers=self.headers)
        if response.status_code == 200:
            item_id = response.json()['id']
            self.remote_path_to_id[path] = item_id
            return item_id
        return None

    def _upload_file(self, local_path: str, rel_path: str):
        self.check_stop()
        target_path = f"{self.remote_path}/{rel_path}" if self.remote_path else rel_path
        
        if self.dry_run:
            self.logger(f"   [DRY RUN] UPLOAD: {rel_path}")
            return True

        # Graph API Upload (PUT)
        # For larger files, an upload session is needed. This handles small files (<4MB).
        url = f"{GRAPH_API_ENDPOINT}/me/drive/root:/{target_path}:/content"
        
        try:
            with open(local_path, 'rb') as f:
                response = requests.put(url, headers=self.headers, data=f)
            
            if response.status_code in [200, 201]:
                return True
            self.logger(f"❌ Error uploading {rel_path}: {response.text}")
            return False
        except Exception as e:
            self.logger(f"❌ Upload exception {rel_path}: {e}")
            return False

    def _create_remote_folder(self, rel_path: str):
        self.check_stop()
        # Determine parent path and name
        parts = rel_path.split('/')
        folder_name = parts[-1]
        parent_rel = "/".join(parts[:-1])
        
        full_parent_path = f"{self.remote_path}/{parent_rel}" if parent_rel else self.remote_path
        if not full_parent_path: full_parent_path = "" # root

        if self.dry_run:
            self.logger(f"   [DRY RUN] CREATE FOLDER: {rel_path}")
            return True

        parent_id = self._resolve_path_to_id(full_parent_path)
        if not parent_id:
            # If parent doesn't exist, we need to create it recursively
            # For simplicity, we assume standard recursive creation order from main loop handles this,
            # or we fail if deep nesting creation order is wrong.
            # In the execute_sync loop, we iterate ensuring parents are processed first.
            self.logger(f"❌ Cannot create {rel_path}, parent not found.")
            return False

        url = f"{GRAPH_API_ENDPOINT}/me/drive/items/{parent_id}/children"
        body = {
            "name": folder_name,
            "folder": {},
            "@microsoft.graph.conflictBehavior": "replace"
        }
        response = requests.post(url, headers=self.headers, json=body)
        if response.status_code == 201:
            new_id = response.json()['id']
            full_path = f"{self.remote_path}/{rel_path}" if self.remote_path else rel_path
            self.remote_path_to_id[full_path] = new_id
            return True
        self.logger(f"❌ Error creating folder {rel_path}: {response.text}")
        return False

    def _download_file(self, item_id: str, local_dest: str):
        self.check_stop()
        if self.dry_run:
            self.logger(f"   [DRY RUN] DOWNLOAD: {item_id} -> {os.path.basename(local_dest)}")
            return True

        os.makedirs(os.path.dirname(local_dest), exist_ok=True)
        url = f"{GRAPH_API_ENDPOINT}/me/drive/items/{item_id}/content"
        
        try:
            response = requests.get(url, headers=self.headers, stream=True)
            if response.status_code == 200:
                with open(local_dest, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=8192):
                        f.write(chunk)
                return True
            self.logger(f"❌ Error downloading {item_id}: {response.text}")
            return False
        except Exception as e:
            self.logger(f"❌ Download exception: {e}")
            return False

    def _delete_local(self, local_path: str):
        self.check_stop()
        if not os.path.exists(local_path): return True
        if self.dry_run:
            self.logger(f"   [DRY RUN] DELETE LOCAL: {local_path}")
            return True
        try:
            if os.path.isdir(local_path): shutil.rmtree(local_path)
            else: os.remove(local_path)
            return True
        except Exception as e:
            self.logger(f"❌ Error deleting local {local_path}: {e}")
            return False

    def _delete_remote(self, item_id: str, rel_path: str):
        self.check_stop()
        if self.dry_run:
            self.logger(f"   [DRY RUN] DELETE REMOTE: {rel_path}")
            return True
        
        url = f"{GRAPH_API_ENDPOINT}/me/drive/items/{item_id}"
        response = requests.delete(url, headers=self.headers)
        if response.status_code == 204:
            return True
        self.logger(f"❌ Error deleting remote {rel_path}: {response.text}")
        return False

    def execute_sync(self) -> tuple[bool, str]:
        try:
            self._authenticate()
            self.check_stop()

            if not os.path.exists(self.local_path):
                return (False, f"Local path '{self.local_path}' does not exist.")

            # 1. Scan
            self.logger("📋 Scanning local and remote files...")
            local_items = self._get_local_files_map()
            remote_items = self._get_remote_files_map()
            
            self.logger(f"   Found {len(local_items)} local items.")
            self.logger(f"   Found {len(remote_items)} remote items.")

            # 2. Sync
            items_uploaded = 0
            items_downloaded = 0
            items_deleted_local = 0
            items_deleted_remote = 0
            items_skipped = 0
            items_ignored = 0
            
            items_skipped_remote = remote_items.copy()

            # Process Local
            for rel_path, local_data in local_items.items():
                self.check_stop()
                
                if local_data['is_folder']:
                    if rel_path in items_skipped_remote:
                        items_skipped_remote.pop(rel_path)
                    else:
                        # Folder local only
                        if self.action_local == "upload":
                            if self._create_remote_folder(rel_path): items_uploaded += 1
                        elif self.action_local == "delete_local":
                            self.logger(f"   DELETING LOCAL FOLDER: {rel_path}")
                            if self._delete_local(local_data['path']): items_deleted_local += 1
                        elif self.action_local == "ignore_local":
                            items_ignored += 1
                    continue

                remote_data = items_skipped_remote.get(rel_path)
                if remote_data:
                    items_skipped += 1
                    items_skipped_remote.pop(rel_path)
                else:
                    # Local orphan
                    if self.action_local == "upload":
                        self.logger(f"   UPLOADING: {rel_path}")
                        if self._upload_file(local_data['path'], rel_path): items_uploaded += 1
                    elif self.action_local == "delete_local":
                        self.logger(f"   DELETING LOCAL: {rel_path}")
                        if self._delete_local(local_data['path']): items_deleted_local += 1
                    elif self.action_local == "ignore_local":
                        self.logger(f"   IGNORING LOCAL: {rel_path}")
                        items_ignored += 1

            # Process Remote
            for rel_path, remote_data in items_skipped_remote.items():
                self.check_stop()
                
                if remote_data['is_folder']:
                    if self.action_remote == "delete_remote":
                        self.logger(f"   DELETING REMOTE FOLDER: {rel_path}")
                        if self._delete_remote(remote_data['id'], rel_path): items_deleted_remote += 1
                    continue

                if self.action_remote == "download":
                    self.logger(f"   DOWNLOADING: {rel_path}")
                    local_dest = os.path.join(self.local_path, rel_path)
                    if self._download_file(remote_data['id'], local_dest): items_downloaded += 1
                elif self.action_remote == "delete_remote":
                    self.logger(f"   DELETING REMOTE: {rel_path}")
                    if self._delete_remote(remote_data['id'], rel_path): items_deleted_remote += 1
                elif self.action_remote == "ignore_remote":
                    self.logger(f"   IGNORING REMOTE: {rel_path}")
                    items_ignored += 1

            total = items_uploaded + items_downloaded + items_deleted_local + items_deleted_remote
            
            self.logger("\n--- Sync Execution Summary ---")
            if total == 0 and items_ignored == 0:
                msg = "No changes needed."
            else:
                prefix = "Simulated" if self.dry_run else "Completed"
                msg = f"{prefix} {total} actions. (Up: {items_uploaded}, Down: {items_downloaded}, Del-L: {items_deleted_local}, Del-R: {items_deleted_remote})"
            
            self.logger(f"✅ {msg}")
            return (True, msg)

        except Exception as e:
            self.logger(f"❌ Error: {e}")
            return (False, str(e))

File: backend/src/web/OneDriveSync.cpp
#include "OneDriveSync.hpp"

const QString GRAPH_ENDPOINT = "https://graph.microsoft.com/v1.0";

OneDriveSync::OneDriveSync(const QString &localPath, const QString &remoteFolderName,
                           const QString &clientId, const QString &token,
                           bool dryRun, Logger logger,
                           const QString &actionLocal, const QString &actionRemote,
                           QObject *parent)
    : AbstractSync(localPath, remoteFolderName, dryRun, logger, actionLocal, actionRemote, parent),
      m_accessToken(token) {
    m_nam = new QNetworkAccessManager(this);
}

void OneDriveSync::executeSync() {
    try {
        checkStop();
        if (m_accessToken.isEmpty()) throw std::runtime_error("Missing Access Token");

        log("📋 Scanning local and remote files...");
        auto localItems = scanLocalFiles();
        auto remoteItems = scanRemoteFiles(); // Implementation below

        log(QString("   Found %1 local items, %2 remote items.").arg(localItems.size()).arg(remoteItems.size()));

        QMap<QString, FileMetadata> remoteSkipped = remoteItems;
        int up = 0, down = 0, delL = 0, delR = 0;

        // 1. Process Local
        for (auto it = localItems.begin(); it != localItems.end(); ++it) {
            checkStop();
            QString relPath = it.key();
            
            if (it->isFolder) {
                if (remoteSkipped.contains(relPath)) {
                    remoteSkipped.remove(relPath);
                } else if (m_actionLocal == "upload") {
                    if (createRemoteFolder(relPath)) up++;
                } else if (m_actionLocal == "delete_local") {
                    if (deleteLocal(it->path)) delL++;
                }
                continue;
            }

            if (remoteSkipped.contains(relPath)) {
                remoteSkipped.remove(relPath); // Exists in both
            } else {
                if (m_actionLocal == "upload") {
                    if (uploadFile(it->path, relPath)) up++;
                } else if (m_actionLocal == "delete_local") {
                    if (deleteLocal(it->path)) delL++;
                }
            }
        }

        // 2. Process Remote Orphans
        for (auto it = remoteSkipped.begin(); it != remoteSkipped.end(); ++it) {
            checkStop();
            QString relPath = it.key();

            if (it->isFolder) {
                if (m_actionRemote == "delete_remote") {
                    if (deleteRemote(it->id, relPath)) delR++;
                }
                continue;
            }

            if (m_actionRemote == "download") {
                QString dest = QDir(m_localPath).filePath(relPath);
                if (downloadFile(it->id, dest)) down++;
            } else if (m_actionRemote == "delete_remote") {
                if (deleteRemote(it->id, relPath)) delR++;
            }
        }
        
        log(QString("✅ Sync Complete. Up: %1, Down: %2, Del-L: %3, Del-R: %4").arg(up).arg(down).arg(delL).arg(delR));

    } catch (const std::exception &e) {
        log(QString("❌ Error: %1").arg(e.what()));
    }
}

// --- API Implementation ---

QByteArray OneDriveSync::makeRequest(const QString &verb, const QString &url, 
                                     const QByteArray &data, const QString &contentType, int *statusCode) {
    QNetworkRequest req(QUrl(url));
    req.setRawHeader("Authorization", ("Bearer " + m_accessToken).toUtf8());
    if (!data.isEmpty() || verb == "POST" || verb == "PUT") {
        req.setHeader(QNetworkRequest::ContentTypeHeader, contentType);
    }

    QNetworkReply *reply = nullptr;
    if (verb == "GET") reply = m_nam->get(req);
    else if (verb == "POST") reply = m_nam->post(req, data);
    else if (verb == "PUT") reply = m_nam->put(req, data);
    else if (verb == "DELETE") reply = m_nam->deleteResource(req);

    QEventLoop loop;
    connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    loop.exec();

    if (statusCode) *statusCode = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
    
    QByteArray result = reply->readAll();
    bool error = reply->error() != QNetworkReply::NoError;
    reply->deleteLater();

    if (error) {
        // Simple error handling, in real app parse JSON error
        log("API Error: " + url + " | " + result); 
    }
    return result;
}

QString OneDriveSync::resolvePathToId(const QString &path) {
    if (path.isEmpty()) return "root";
    if (m_remotePathToId.contains(path)) return m_remotePathToId[path];

    // Graph API allows path addressing: /drive/root:/path/to/folder
    QString url = GRAPH_ENDPOINT + "/me/drive/root:/" + path;
    int code;
    QByteArray resp = makeRequest("GET", url, {}, "", &code);
    if (code == 200) {
        QJsonObject obj = QJsonDocument::fromJson(resp).object();
        QString id = obj["id"].toString();
        m_remotePathToId[path] = id;
        return id;
    }
    return "";
}

QMap<QString, FileMetadata> OneDriveSync::scanRemoteFiles() {
    QMap<QString, FileMetadata> items;
    QString rootId = resolvePathToId(m_remotePath);
    if (rootId.isEmpty()) return items;

    struct QueueItem { QString id; QString relPath; };
    QList<QueueItem> queue;
    queue.append({rootId, ""});

    while (!queue.isEmpty()) {
        checkStop();
        QueueItem current = queue.takeFirst();
        QString url = GRAPH_ENDPOINT + "/me/drive/items/" + current.id + "/children";

        while (!url.isEmpty()) {
            QByteArray json = makeRequest("GET", url);
            QJsonObject root = QJsonDocument::fromJson(json).object();
            QJsonArray vals = root["value"].toArray();

            for (const auto &v : vals) {
                QJsonObject item = v.toObject();
                QString name = item["name"].toString();
                QString id = item["id"].toString();
                bool isFolder = item.contains("folder");
                QString relPath = current.relPath.isEmpty() ? name : current.relPath + "/" + name;

                FileMetadata meta;
                meta.id = id;
                meta.isFolder = isFolder;
                items.insert(relPath, meta);

                if (isFolder) {
                    queue.append({id, relPath});
                    QString fullPathKey = m_remotePath.isEmpty() ? relPath : m_remotePath + "/" + relPath;
                    m_remotePathToId[fullPathKey] = id;
                }
            }
            // Handle Pagination
            if (root.contains("@odata.nextLink")) url = root["@odata.nextLink"].toString();
            else url = "";
        }
    }
    return items;
}

bool OneDriveSync::uploadFile(const QString &localPath, const QString &relPath) {
    checkStop();
    if (m_dryRun) { log("   [DRY RUN] UPLOAD: " + relPath); return true; }

    QString target = m_remotePath.isEmpty() ? relPath : m_remotePath + "/" + relPath;
    QString url = GRAPH_ENDPOINT + "/me/drive/root:/" + target + ":/content";
    
    QFile f(localPath);
    if (!f.open(QIODevice::ReadOnly)) return false;
    QByteArray content = f.readAll();
    
    int code;
    makeRequest("PUT", url, content, "application/octet-stream", &code);
    return (code == 200 || code == 201);
}

bool OneDriveSync::createRemoteFolder(const QString &relPath) {
    checkStop();
    if (m_dryRun) { log("   [DRY RUN] CREATE FOLDER: " + relPath); return true; }

    // Logic: Find parent ID, create child. 
    // Simplified: Using Path addressing for creation is tricky if parent doesn't exist.
    // Assuming hierarchical traversal creates parents first.
    int lastSlash = relPath.lastIndexOf('/');
    QString parentRel = (lastSlash == -1) ? "" : relPath.left(lastSlash);
    QString name = (lastSlash == -1) ? relPath : relPath.mid(lastSlash + 1);
    
    QString fullParent = m_remotePath.isEmpty() ? parentRel : m_remotePath + "/" + parentRel;
    if (fullParent.endsWith('/')) fullParent.chop(1);
    
    QString parentId = resolvePathToId(fullParent);
    if (parentId.isEmpty()) return false;

    QString url = GRAPH_ENDPOINT + "/me/drive/items/" + parentId + "/children";
    QJsonObject body;
    body["name"] = name;
    body["folder"] = QJsonObject();
    body["@microsoft.graph.conflictBehavior"] = "replace";
    
    int code;
    QByteArray resp = makeRequest("POST", url, QJsonDocument(body).toJson(), "application/json", &code);
    if (code == 201) {
        QJsonObject res = QJsonDocument::fromJson(resp).object();
        QString fullKey = m_remotePath.isEmpty() ? relPath : m_remotePath + "/" + relPath;
        m_remotePathToId[fullKey] = res["id"].toString();
        return true;
    }
    return false;
}

bool OneDriveSync::downloadFile(const QString &itemId, const QString &localDest) {
    checkStop();
    if (m_dryRun) { log("   [DRY RUN] DOWNLOAD: " + itemId); return true; }

    QDir().mkpath(QFileInfo(localDest).absolutePath());
    QString url = GRAPH_ENDPOINT + "/me/drive/items/" + itemId + "/content";
    QByteArray data = makeRequest("GET", url);
    
    QFile f(localDest);
    if (f.open(QIODevice::WriteOnly)) {
        f.write(data);
        f.close();
        return true;
    }
    return false;
}

bool OneDriveSync::deleteRemote(const QString &itemId, const QString &relPath) {
    checkStop();
    if (m_dryRun) { log("   [DRY RUN] DELETE REMOTE: " + relPath); return true; }
    
    QString url = GRAPH_ENDPOINT + "/me/drive/items/" + itemId;
    int code;
    makeRequest("DELETE", url, {}, "", &code);
    return (code == 204);
}